src/formulation_definitions.jl

```diff
@@ -0,0 +1,295 @@
+        return new(max_active_power)
+    end
+end
+
+"""
+    HydroEnergyReservoirFormulation <: GeneratorStorageFormulation
+
+# Arguments
+$(TYPEDFIELDS)
+"""
+struct HydroEnergyReservoirFormulation <: GeneratorStorageFormulation
+    "Name of time series to use for max active power"
+    max_active_power::String
+    "Name of time series to use for inflow"
+    inflow::String
+    "Name of time series to use for storage capacity"
+    storage_capacity::String
+
+    function HydroEnergyReservoirFormulation(;
+        max_active_power="max_active_power",
+        inflow="inflow",
+        storage_capacity="storage_capacity",
+    )
+        return new(max_active_power, inflow, storage_capacity)
+    end
+end
+
+"""
+    StorageFormulation <: AbstractPRASFormulation
+
+Objects in Sienna that behave like storage are mapped to storage in PRAS.
+"""
+struct StorageFormulation <: AbstractPRASFormulation end
+
+"""
+    DevicePRASModel{D <: PSY.Device, B <: AbstractPRASFormulation}
+
+# Arguments
+
+- D <: PSY.Device: Device type
+$(TYPEDFIELDS)
+
+A DevicePRASModel, like a DeviceModel in PowerSimulations, assigns a type of Component
+to a specific formulation. Unlike Sienna, we put configuration information
+in the formulation itself.
+"""
+struct DevicePRASModel{D <: PSY.Device, B <: AbstractPRASFormulation}
+    "Formulation containing configuration"
+    formulation::B
+
+    function DevicePRASModel(
+        ::Type{D},
+        formulation::B,
+    ) where {D <: PSY.Device, B <: AbstractPRASFormulation}
+        return new{D, B}(formulation)
+    end
+end
+
+"""
+    $(TYPEDSIGNATURES)
+    
+# Arguments
+- `::Type{D}`: Device type
+- `::Type{B}`: Formulation type
+- `time_series_names::Dict{Symbol, String}`: Mapping of time series `Symbol` to names
+- `kwargs...`: Additional arguments to pass to the formulation constructor
+
+Keyword arguments in DevicePRASModel are passed to the
+formulation constructor.
+
+You may also pass a `time_series_names` Dict to map time series `Symbol` to names.
+
+# Example
+
+```julia
+DevicePRASModel(
+    PSY.Generator,
+    GeneratorFormulation(max_active_power="max_active_power"),
+)
+```
+
+```julia
+DevicePRASModel(
+    PSY.HydroEnergyReservoir,
+    HydroEnergyReservoirFormulation;
+    max_active_power="max_active_power",
+    inflow="inflow",
+    storage_capacity="storage_capacity",
+)
+```
+
+```julia
+DevicePRASModel(
+    PSY.HybridSystem,
+    HybridSystemFormulation;
+    time_series_names=Dict(:max_active_power="max_active_power"),
+)
+```
+"""
+function DevicePRASModel(
+    ::Type{D},
+    ::Type{B};
+    time_series_names::Dict{Symbol, String}=Dict{Symbol, String}(),
+    kwargs...,
+) where {D <: PSY.Device, B <: AbstractPRASFormulation}
+    formulation = B(; time_series_names..., kwargs...)
+    return DevicePRASModel(D, formulation)
+end
+
+"""
+Check whether a DevicePRASModel applies to a given type
+"""
+function appliestodevice(::DevicePRASModel{D}, ::Type{T}) where {D, T}
```

Comment:
I'm not sure I follow how this doesn't allow _PSY.EnergyReservoirStorage_ to not have a _GeneratorFormulation_?

Reply:
You can totally do that in this framework. This only tells you about the `PSY.Device` subtype `D` and has nothing about the formulation type itself. Unless we specify some relation between the type parameters in DevicePRASModel, then it will error when it tries to actually apply a model.